### ISSUE TITLE
add a docker-compose.yml

### ISSUE THREAD

**Issue by zhenyulin** [xygon/react-redux-typescript-dotnet-core#1]
Hi @xygon , I'm currently looking into this repo in details, can you help to set up a Dockerfile or docker-compose.yml for development over docker?

**Comment by xygon** [comment]
I can, but why do you think it is needed?

**Comment by xygon** [comment]
@zhenyulin ^^^^^

**Comment by zhenyulin** [comment]
@xygon it would be very helpful and much faster to start the development in this way, which could probably take me a day or two to set up the environment just to get the repo running.

**Comment by xygon** [comment]
@zhenyulin -the .net part is using asp.net coer so should run on a mac.

**Comment by zhenyulin** [comment]
@xygon the version management can be a bit messy here, and I'm inclining to have a consistent setup to minimise unexpected results. Also this will be needed to further inspecting things in the CI environment. Can you do a quick Dockerfile?

**Comment by xygon** [comment]
@zhenyulin I'm on a train and  the connection isn't great so getting the base image isn't really feasable at the moment. Also, this is supposed to be about looking at a different type of toolset and evaluationg it - it isn't about being consistent with anything else or putting it in any sort of CI environment. All you need is .net core and it should just run (I haven't go a mac, so haven't tested that, but it works on my windows machines and the linux virtual machines)

**Comment by zhenyulin** [comment]
@xygon by toolset, I suppose we include code quality check tools(linting, unit test, static check?), development tools(possibly docker-compose.yml) here? The code quality check tools will need to be tested in the CI environment as well the local development environment. 

I'll further look into set this up natively on Mac. Let me know if you get a chance to make a Dockerfile afterwards.

**Comment by xygon** [comment]
@zhenyulin by toolset I just meant typescript / react / redux and a restful api in one project. Then adding whatever is needed to do the *Planed changes* from thge README. At the moment I really don't care about the CI environment - we may at some point but not yet!

**Comment by zhenyulin** [comment]
@xygon just to confirm, what's the major goals we're trying to achieve here? I have got a boilderplate setup with the only difference being the backend running on purely nodejs without dotnet. Would be great to discuss pros and cons of those two setup.

**Comment by xygon** [comment]
@zhenyulin  The goal is to do the *Planed changes* in this repo to see how it works. It is a learning excersie to see if the toolset / technology fits what we want to do. We are not intending to use nodejs for the api, so that wouldn't be appropriate. Also, how have you got server side rendering working with this project without dotnet? It isn't that nodejs couldn't be used, it definately could, but taht isn't the goasl of this piece of work.

**Comment by zhenyulin** [comment]
@xygon server side rendering has been setup https://github.com/finbourne/website/blob/master/server/render.js
but currently disabled, as it is slower than the client side rendering experience, when it comes to router changes. Would be useful to inspect if it provides more benefit on the product website. For the API I think we're going to build a typescript SDK?

**Comment by xygon** [comment]
@zhenyulin I mean server side rendering for this projrct *react-redux-typescript-dotnet-core*. What we are doing here does not, and must not, depend on anything else that has already been done for finbourne - this is just looking at a different toolset and evaluating it. If we go ahead with this toolset then we would develop a typescript SDK but for this project we won't have one.

**Comment by zhenyulin** [comment]
@xygon I'm a bit confused here, if we're building another more specific set of APIs on Dotnet, it doesn't seem the typescript SDK would be very necessary, while if we're using the current generic API, then the SDK can be built to get the specific data for the UI.

**Comment by xygon** [comment]
@zhenyulin to be clear, all Im looking for is some pull requests to satisfy the *Planned changes* in the README - nothing else. They should be the minimum required to realistically achieve the change. If after doing that we decide that this is the way to proceed then we can add further tooling and / or consider other technologies.

**Comment by zhenyulin** [comment]
@xygon I assume we're testing this setup to see how it fits our product development, then tooling and pipeline are factors need to be considered as well in order to achieve consistent code style

**Comment by zhenyulin** [comment]
@xygon what exact aspects we're looking at here by those changes? It seems a bit messy to put dotnet unit test together with javascript unit test

**Comment by xygon** [comment]
@zhenyulin how it fits into our production environment will only become a factor if we decide that this is thge toolset we are going to use. We arwe a small company - we don't need tooling for things like consistent code style - we should just talk to each other.

**Comment by xygon** [comment]
@zhenyulin these are tests for the seb site - which would be comprised of both client side and server side code. The server side code in this case is not the finbourne pulic API, it is web site specific - and should be coupled to the web site. In that context I thionk it is correct to test all aspects of the site - javascript and c#.

**Comment by zhenyulin** [comment]
@xygon they don't necessarily need to be coupled in one repo in order to be tested, and this may create extra complexity for setting up the CI and optimise the developer experience

**Comment by xygon** [comment]
@zhenyulin the dotnet tests would be in a seperate project in the same solution, as that is the standard way of testing dotnet. The javascript tests would be in *spec* files beside the things they tests, as that is the javascript way of doing it.

**Comment by xygon** [comment]
@zhenyulin the point is I want them to be coupled - they serve no purpose in isolation. The server side code in this context is not supposed to be generic,  it is specific to the web site and would contain logic that is web site specific. This is very different to talking to code talking to a generic API that could be serving multiple clients.

**Comment by xygon** [comment]
@zhenyulin for example, what a user can do on the web site doesn't belong in the generic finbourne API - it is web site specific. Another example would be a users prefered settings for the site.

**Comment by zhenyulin** [comment]
@xygon yep, but that doesn't necessary require them to be coupled, they can be in different repo and use docker-compose for local development; this has huge benefit if we would like to open the specific API to the public or the clients, as it wouldn't require us to decouple them again.

**Comment by xygon** [comment]
@zhenyulin as I said, I want them to be coupled. I see nothing but down side to seperating them. You would never expose a site-specific API as a *public* API - that is the point of what I am saying. If any of the logic in the site-specific server side code becomes generic then that logic may be moved to a public API, but I have never had this happen. The whole point is that the site logic is shared between the client and server side code. The server side code is not suppose to be re-usable outside the site. Thw public API would never be in this solution - this is web site specific.

**Comment by xygon** [comment]
@zhenyulin what technology to use for the server side *site specific* code is a different question, but in this case it is dotnet.

**Comment by zhenyulin** [comment]
@xygon this would be a question worth discussing with the whole team, considering factors of requirements of this part of the API, later client business model, deployment pipeline, request and scalability, etc.

**Comment by zhenyulin** [comment]
I've created a working Dockerfile based on dotnet:1.1.2-sdk, should be OK to use that making the docker-compose.yml for development; I need the access to push changes to the repo.

**Comment by zhenyulin** [comment]
@xygon I had a quick look into the requirements for the site specific API, which would majorly just be the personalisation settings, and can actually be done in Node.js with typescript in a light weight and robust manner

**Comment by xygon** [comment]
I know we could do it in nodejs, I covered that in the earlier comments.

The point is to try what is in the project and see how it goes – not to come up with as many alternatives as possible!

Also, there are no known requirements for the site specific API, the things I mentioned earlier are just some examples that I’ve had to do in the past. The actual requirements would be site specific – that’s the point.

**Comment by zhenyulin** [comment]
@xygon I need to admit that it is a relatively large leap here, going from Javascript to C# directly, which doesn't seem to be very efficient in terms of building website specific API, as they'll be relatively light weight. And it looks in the current solution it is running a JavaScript runtime on Dotnet core, so I think starting with Node.js in typescript would be a better solution here. What do you think?

**Comment by xygon** [comment]
The point is to actually give this a try and see how it goes. There are lots of choices in how to do this, and this is one of them.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by zhenyulin** [comment]
I also had a quick take on what we want to achieve via this solution on Node.js
https://github.com/zhenyulin/ultimate-hot-boilerplate/tree/typescript
